index.js

```diff
@@ -109,6 +169,266 @@ export const immediate = x => Future ((rej, res) => {
+  default: return reject (new Error (`Unsupported protocol '${protocol}'`));
+  }
+};
+
+//# request :: Object -> Url -> Readable Buffer -> Future Error IncomingMessage
+//.
+//. Given an Object of [http options][], a String containing the request URL,
+//. and a [Readable][] stream of [Buffer][]s to be sent as the request body,
+//. returns a Future which makes an HTTP request and resolves with its
+//. [IncomingMessage][]. If the Future is cancelled, the request is aborted.
+//.
+//. This is the "lowest level" function for making HTTP requests. It does not
+//. handle buffering, encoding, content negotiation, or anything really.
+//. For most use-cases, you can use one of the more specialized functions:
+//.
+//. * [`send`](#sent): Make a generic HTTP request.
+//. * [`retrieve`](#retrieve): Make a GET request.
+export const request = options => url => body => {
+  const location = new URL (url);
+  return chain (agent => Future ((rej, res) => {
+    const req = agent.request (location, options, res);
+    pipeline (body, req, e => e && rej (e));
+    return () => {
+      req.abort ();
+    };
+  })) (getAgent (location.protocol));
+};
+
+//# messageMeta :: IncomingMessage -> MessageMeta
+//.
+//. Extracts the common meta-data from an [IncomingMessage][].
+//.
+//. The `MessageMeta` type is defined as
+//. `{ code :: Number, message :: String, headers :: StrMap String }`:
+//.
+//. * `code`: The HTTP status code.
+//. * `message`: The HTTP status message.
+//. * `headers`: A StrMap of the HTTP headers.
+export const messageMeta = message => ({
+  code: message.statusCode,
+  message: message.statusMessage,
+  headers: message.headers,
+});
+
+//# bufferResponse :: Charset -> IncomingMessage -> Future Error {meta :: MessageMeta, body :: String}
+//.
+//. A version of [`buffer`](#buffer) specialized in [IncomingMessage][]s.
+//.
+//. Given a charset and an IncomingMessage, returns a Future with the
+//. [MessageMeta](#messagemeta) and the buffered, encoded, message body.
+export const bufferResponse = charset => message => {
+  const meta = messageMeta (message);
+  const eventualBody = bufferString (charset) (message);
+  return map (body => ({meta, body})) (eventualBody);
+};
+
+// defaultCharset :: String
+const defaultCharset = 'utf8';
+
+// defaultContentType :: String
+const defaultContentType = 'text/plain; charset=' + defaultCharset;
+
+// charsetRegex :: RegExp
+const charsetRegex = /charset=([^;\s]+)/;
+
+//# autoBufferResponse :: IncomingMessage -> Future Error {meta :: MessageMeta, body :: String}
+//.
+//. Given an IncomingMessage, buffers and decodes the message body using the
+//. charset provided in the message headers. Falls back to utf8 if the
+//. charset was not provided.
+//.
+//. Returns a Future with the [MessageMeta](#messagemeta) and the buffered,
+//. encoded, message body.
+export const autoBufferResponse = message => {
+  const contentType = message.headers['content-type'] || defaultContentType;
```

Comment:
Are you intentionally treating `''` as `undefined` here?

Reply:
It doesn't matter because the code paths converge on the same result in the following statements. :man_shrugging: